### ISSUE TITLE
Should cascade delete work when specifying custom class for join entity

### ISSUE THREAD

**Issue by roblangston** [dotnet/efcore#34136]
I have what is a simple database schema .  It contains two entities, contact, address and a link table contactaddress.

**Database Tables:-**

```
Contact
(
      Id int,
      Name string
)

Address
(
    Id int,
    Address string
)

ContactAddress
(
   Id int,
   ContactId,
   AddressId
)
```

**C# Classes**

```
Contact
(
    Id int,
    List<Address> Addresses
)

Address
(
    Id int,
    Address string
)

ContactAddress
(
   Id int,
   ContactId,
   AddressId
)
```

I have overridden the OnModelCreating method as follows:-

```
protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.UseCollation("NOCASE");

            modelBuilder.Entity<Contact>()
                .HasMany(e => e.Addresses)
                .WithMany(e => e.Contacts)
                .UsingEntity<ContactAddress>(
                l => l.HasOne<Address>().WithMany().HasForeignKey(e => e.AddressId),
                r => r.HasOne<Contact>().WithMany().HasForeignKey(e => e.ContactId));
```

When trying to delete a contact, I was expecting any entries in the ContactAddress table to be deleted (cascade delete) where the ContactId matches contact being deleted.  I understand that any orphaned addresses in the Address table would need to be removed manually if they are not assigned to any other contacts.

This is the delete code:-

```
 var contacts = context.Contacts
                                        .Include(c => c.Addresses)
                                        .Where(c => c.Id == id)
                                        .ExecuteDelete();
```

However it fails with the following error:-

```
Exception thrown: 'System.InvalidOperationException' in Microsoft.EntityFrameworkCore.Relational.dll
An exception of type 'System.InvalidOperationException' occurred in Microsoft.EntityFrameworkCore.Relational.dll but was not handled in user code
The LINQ expression 'DbSet<Contact>()
    .Where(c => c.Id == __id_0)
    .Select(c => IncludeExpression(
        EntityExpression:
        c, 
        NavigationExpression:
        MaterializeCollectionNavigation(
            Navigation: Contact.Addresses,
            subquery: DbSet<ContactAddress>()
                .Where(c0 => EF.Property<int?>(c, "Id") != null && object.Equals(
                    objA: (object)EF.Property<int?>(c, "Id"), 
                    objB: (object)EF.Property<int?>(c0, "ContactId")))
                .Join(
                    inner: DbSet<Address>(), 
                    outerKeySelector: c0 => (object)EF.Property<int?>(c0, "AddressId"), 
                    innerKeySelector: a => (object)EF.Property<int?>(a, "Id"), 
                    resultSelector: (c0, a) => new TransparentIdentifier<ContactAddress, Address>(
                        Outer = c0, 
                        Inner = a
                    ))
                .Select(ti => IncludeExpandingExpressionVisitor.FetchJoinEntity<ContactAddress, Address>(
                    joinEntity: ti.Outer, 
                    targetEntity: ti.Inner))), Addresses)
    )
    .ExecuteDelete()' could not be translated. Additional information: The operation 'ExecuteDelete' requires an entity type which corresponds to the database table to be modified. The current operation is being applied on a non-entity projection. Remove any projection to non-entity types. See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.

```

Am I missing something obvious?  Reading the documentation it implies this scenario should work without having to manually delete child entries before the parent.

**Comment by ajcvickers** [comment]
@roblangston If you are using EF to create the database for you, then it will setup cascade delete in the database. The call to `ExecuteUpdate` should therefore work if you remove the `Include`. Includes are not allowed when using `ExecuteUpdate` because it implies `ExecuteUpdate` can work against multiple tables, which is not possible because the SQL targets a specific table to update.

**Comment by roblangston** [comment]
@ajcvickers Many thanks for the reply.  The SQLite database was created before introducing entity framework.  I revisited the schema and the foreign key definitions were missing the ON DELETE CASCADE value.  Adding these into the database and adjusting the call to OnDelete() as follows results in everything working as expected:-

```
                var contact = context.Contacts
                                        .Where(c => c.Id == id)
                                        .ExecuteDelete();
```
I think where I went wrong was the expectation that EF Core would generate the deletes for each child dependency rather than relying on the DELETE CASCADE as specified against the foreign key on the child tables.  I will close this question, many thanks.
